JSQMessagesViewController/Views/JSQMessagesComposerTextView.m

```diff
@@ -225,18 +213,52 @@ - (NSDictionary *)jsq_placeholderTextAttributes
 
 #pragma mark - UIMenuController
 
-- (BOOL)canBecomeFirstResponder
-{
-    return [super canBecomeFirstResponder];
+- (BOOL)isCustomMenuItemSelector:(SEL)aSelector {
+    for (UIMenuItem *item in self.customMenuItems) {
+        if (item.action == aSelector) {
+            return YES;
+        }
+    }
+    return NO;
 }
 
-- (BOOL)becomeFirstResponder
-{
-    return [super becomeFirstResponder];
+- (NSMethodSignature *)methodSignatureForSelector:(SEL)aSelector {
+    
+    if ([super methodSignatureForSelector:aSelector]) {
+        return [super methodSignatureForSelector:aSelector];
+    }
+    return [self.customTarget methodSignatureForSelector:aSelector];
+}
+
+- (void)forwardInvocation:(NSInvocation *)anInvocation {
+    
+    if ([self isCustomMenuItemSelector:anInvocation.selector]) {
+        [anInvocation invokeWithTarget:self.customTarget];
+    } else {
+        [super forwardInvocation: anInvocation];
+    }
 }
 
 - (BOOL)canPerformAction:(SEL)action withSender:(id)sender {
-    [UIMenuController sharedMenuController].menuItems = nil;
+    if (!self.selectedTextRange.empty) {
+        if (UIMenuController.sharedMenuController.menuItems == nil) {
+            [UIMenuController.sharedMenuController setMenuItems:self.customMenuItems];
+        }
+        if ([self isCustomMenuItemSelector:action]) {
+            return YES;
+        }
+    } else {
+        [UIMenuController.sharedMenuController setMenuItems:nil];
+    }
+    
     return [super canPerformAction:action withSender:sender];
 }
+
+- (void)setCustomMenuItemsForSelectedText:(NSArray<UIMenuItem *> *)menuItems actionsTarget:(id)target {
```

Comment:
Maybe `setCustomMenuItemsForTextSelection`? selected text may indicate it's only for currently selected text

Reply:
And this is true. Custom menu items are only visible for current selected text. No current selected text -> no custom buttons in menu.